src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -31,6 +31,56 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region InodeTracker
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<(UInt64, UInt64), bool>  _visitations;
+
+        /// <summary>
+        /// Construct a new InodeTracker with an initial path
+        /// </summary>
+        internal InodeTracker(string path)
+        {
+            _visitations = new Dictionary<(UInt64, UInt64), bool>();
+            var inodeData = (0UL, 0UL);
+
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out inodeData))
```

Comment:
Minor comment: maybe use the following pattern to avoid `var inodeData = (0UL, 0UL);`?
```CSharp
if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out (UInt64, UInt64) inodeData))
```